### ISSUE TITLE
Visual C++ warning C4268 on H5Opublic.h

### ISSUE THREAD

**Issue by kmccarty-dgi** [HDFGroup/hdf5#3687]
**Describe the bug**
We are trying internally to upgrade our HDF5 version from 1.10.9 to 1.14.2.  In doing so, we see that Visual C++ is now issuing the following warning against the H5Opublic.h header:

```
d:\hdf5-1.14.2\include\H5Opublic.h(1683): warning C4268: 'H5O_TOKEN_UNDEF_g': 'const' static/global data initialized with compiler generated default constructor fills the object with zeros (compiling source file .\h5open.cpp)
```

This is the line in question (preceding two lines quoted for context) --

```
/* The canonical 'undefined' token value */
#define H5O_TOKEN_UNDEF (H5OPEN H5O_TOKEN_UNDEF_g)
H5_DLLVAR const H5O_token_t H5O_TOKEN_UNDEF_g;
```

(I see that these lines don't exist in HDF5 1.10.9 headers.)

**Expected behavior**
I am not clear on whether the warning reflects an actual problem; but if it does, it should be fixed; and if not, could it be determined whether there is any easy way to prevent the warning from being issued?

I haven't been able to find any other discussion of warning C4268 in this specific context on the Internet.  According to Microsoft's docs this warning would only be issued at -W4.

**Platform (please complete the following information)**
 - HDF5 version (if building from a maintenance branch, please include the commit hash)
    *1.14.2*
 - OS and version
    *Windows 10*
 - Compiler and version
    *Visual C++ 19.29.30148 for x64*
 - Build system (e.g. CMake, Autotools) and version
    *Cmake*
 - Any configure options you specified
```
cmake -G "NMake Makefiles" -DCMAKE_BUILD_TYPE=Release ^
    	  -DHDF5_ENABLE_Z_LIB_SUPPORT:BOOL=ON ^
	  -DZLIB_INCLUDE_DIR:PATH=D:\zlib-1.2.12\include ^
	  -DZLIB_LIBRARY:PATH=d:\zlib-1.2.12\lib64\zlib.lib ^
	  -DCURL_INCLUDE_DIR:PATH=D:\curl-8.4.0\include ^
	  -DCURL_LIBRARY:PATH=d:\curl-8.4.0\lib\libcurl_imp.lib ^
  	  -DOPENSSL_ROOT_DIR:PATH=D:\openssl-1.1.1e ^
	  -DHDF5_ENABLE_ROS3_VFD:BOOL=ON ^
	  -DCMAKE_INSTALL_PREFIX:PATH=d:\hdf5-1.14.2 ^
	  -DHDF5_INSTALL_LIB_DIR:PATH=d:\hdf5-1.14.2\lib64 ^
	  -DHDF5_INSTALL_BIN_DIR:PATH=d:\hdf5-1.14.2\bin64 ..
```
 - MPI library and version (parallel HDF5)
    *N/A*

**Additional context**

Visual C++ compiler flags in play:
```
-MP -std:c++20 -EHsc -permissive-   -D_CRT_SECURE_NO_DEPRECATE -D_CRT_NONSTDC_NO_DEPRECATE  -D_USE_MATH_DEFINES -DNOMINMAX -DWIN32_LEAN_AND_MEAN -Dstrcasecmp=_stricmp -Dstrncasecmp=_strnicmp -we4311 -we4473 -we4474 -we4477 -w15038 -MD -O2 -nologo -W4 -wd4100 -wd4127 -wd4522 -wd4706  -wd5033 -wd5054 -wd6326 -wd6993  -D_SCL_SECURE_NO_WARNINGS  -D_HAS_AUTO_PTR_ETC=1  -D_SILENCE_CXX17_ALLOCATOR_VOID_DEPRECATION_WARNING  -D_SILENCE_CXX20_IS_POD_DEPRECATION_WARNING -Zc:strictStrings- -Zc:ternary- -Id:\hdf5-1.14.2\include -ID:\boost_1_83_0  -DBOOST_ALL_NO_LIB  -DBOOST_FILESYSTEM_NO_DEPRECATED  -DBOOST_GEOMETRY_ROBUSTNESS_ALTERNATIVE  -DH5_BUILT_AS_DYNAMIC_LIB  -c -Fowin64\ .\h5open.cpp
```

**Comment by derobins** [comment]
Is the problem that you see this warning when consuming the HDF5 library from a C++ application?

The value of the global variable H5O_TOKEN_UNDEF_g is set in H5Oint.c to all 255 bytes. H5_DLLVAR (defined in H5api_adpt.h) *should* include extern so the compiler realizes the header file contains a declaration, not a definition, preventing assigning a value to it. Perhaps this is getting messed up somehow?

Tokens didn't exist in HDF5 1.10, btw. They were added to generalize file addresses to virtual object layer (VOL) connectors that map HDF5 API calls to non-file storage.

**Comment by kmccarty-dgi** [comment]
> Is the problem that you see this warning when consuming the HDF5 library from a C++ application?

Yes, exactly.

> H5_DLLVAR (defined in H5api_adpt.h) should include extern so the compiler realizes the header file contains a declaration, not a definition, preventing assigning a value to it. Perhaps this is getting messed up somehow?

From what I can see in H5api_adpt.h in 1.14.2, in the case where `H5_BUILT_AS_DYNAMIC_LIB` is defined, `_MSC_VER` is defined and `hdf5_shared_EXPORTS` is not defined -- which I presume is what gets seen by our code that consumes it -- `H5_DLLVAR` gets defined as `__declspec(dllimport)` at line 34 -- no `extern` here.  And since `H5O_TOKEN_UNDEF_g` is declared in H5Opublic.h as being `const` with no explicit `extern`, it implicitly has static linkage.  Unless I'm misunderstanding something!

**Comment by kmccarty-dgi** [comment]
Some research suggests that MSVC has `__declspec(dllimport)` itself imply `extern` ... so I'm not clear on whether the presence of `const` without explicit `extern` overrides that implication and forces `static`; or if the C4268 warning is simply spurious here.  I haven't found anything online that disambiguates between these two possibilities.

**Comment by kmccarty-dgi** [comment]
Some experimentation with a tiny test case leads me to believe that in this case the C4268 warning is spurious, and the consumer of the library will in fact correctly import the value of the variable that was compiled into the dynamic library.  (Though I certainly suggest that you verify this for yourselves in the case of the actual HDF5 code.)

However, it also appears the the warning can be quieted by including an explicit `extern` in the definition of `H5DLL_VAR`.  So I'm happy to leave it up to you folks as to whether it makes more sense to you to close this as wontfix, or to add the explicit `extern` just for the sake of quieting the warning.

Thanks for considering!

**Comment by kmccarty-dgi** [comment]
For the record I've issued a defect report against Visual C++ with my small test case:
https://developercommunity.visualstudio.com/t/Spurious-C4268-warning-issued-against-co/10493767

**Comment by derobins** [comment]
I'll take a look at how `__declspec(dllimport)` is handled, especially in C++. A fix might have to wait for 1.14.4, though. We're just about to release 1.14.3 and I hesitate to make major changes in a hurry just before a release.

**Comment by derobins** [comment]
Microsoft fixed their issue:

https://developercommunity.visualstudio.com/t/Spurious-C4268-warning-issued-against-co/10493767?sort=newest&ftype=problem

Thanks for reporting it to Microsoft!
